File: backend/src/main/java/org/example/backend/domain/aquarium/service/AquariumLogService.java
package org.example.backend.domain.aquarium.service;

import lombok.RequiredArgsConstructor;
import org.example.backend.domain.aquarium.dto.AquariumLogRequestDto;
import org.example.backend.domain.aquarium.dto.AquariumLogResponseDto;
import org.example.backend.domain.aquarium.entity.Aquarium;
import org.example.backend.domain.aquarium.entity.AquariumLog;
import org.example.backend.domain.aquarium.repository.AquariumLogRepository;
import org.example.backend.domain.aquarium.repository.AquariumRepository;
import org.example.backend.domain.log.service.AbstractLogService;
import org.example.backend.global.exception.ErrorCode;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class AquariumLogService extends AbstractLogService<AquariumLog, AquariumLogRequestDto, AquariumLogResponseDto, Aquarium> {

    private final AquariumLogRepository aquariumLogRepository;
    private final AquariumRepository aquariumRepository;

    @Override
    protected JpaRepository<AquariumLog, Long> getLogRepository() {
        return aquariumLogRepository;
    }

    @Override
    protected JpaRepository<Aquarium, Long> getParentRepository() {
        return aquariumRepository;
    }

    @Override
    protected ErrorCode getLogNotFoundErrorCode() {
        return ErrorCode.AQUARIUM_LOG_NOT_FOUND;
    }

    @Override
    protected ErrorCode getParentNotFoundErrorCode() {
        return ErrorCode.AQUARIUM_NOT_FOUND;
    }

    @Override
    protected AquariumLog createEntity(AquariumLogRequestDto requestDto, Aquarium aquarium) {
        return AquariumLog.builder()
                .aquarium(aquarium)
                .temperature(requestDto.getTemperature())
                .ph(requestDto.getPh())
                .logDate(requestDto.getLogDate())
                .build();
    }

    @Override
    protected AquariumLogResponseDto convertToResponseDto(AquariumLog entity) {
        return AquariumLogResponseDto.from(entity);
    }

    @Override
    protected void updateEntity(AquariumLog entity, AquariumLogRequestDto requestDto, Aquarium aquarium) {
        entity.setAquarium(aquarium);
        entity.setTemperature(requestDto.getTemperature());
        entity.setPh(requestDto.getPh());
        entity.setLogDate(requestDto.getLogDate());
    }

    @Override
    protected List<AquariumLog> findByParentId(Long parentId) {
        return aquariumLogRepository.findByAquariumId(parentId);
    }

    // 기존 메서드들을 추상 클래스의 메서드로 위임
    @Transactional
    public AquariumLogResponseDto createLog(AquariumLogRequestDto requestDto) {
        return createLog(requestDto, requestDto.getAquariumId());
    }

    public List<AquariumLogResponseDto> getLogsByAquariumId(Long aquariumId) {
        return getLogsByParentId(aquariumId);
    }

    @Transactional
    public AquariumLogResponseDto updateLog(Long logId, AquariumLogRequestDto requestDto) {
        return updateLog(logId, requestDto, requestDto.getAquariumId());
    }

    @Transactional
    public void deleteLog(Long logId) {
        super.deleteLog(logId);
    }
}

File: backend/src/main/java/org/example/backend/domain/aquarium/service/AquariumLogService.kt
package org.example.backend.domain.aquarium.service

import org.example.backend.domain.aquarium.dto.AquariumLogRequestDto
import org.example.backend.domain.aquarium.dto.AquariumLogResponseDto
import org.example.backend.domain.aquarium.entity.Aquarium
import org.example.backend.domain.aquarium.entity.AquariumLog
import org.example.backend.domain.aquarium.repository.AquariumLogRepository
import org.example.backend.domain.aquarium.repository.AquariumRepository
import org.example.backend.domain.log.service.AbstractLogService
import org.example.backend.global.exception.BusinessException
import org.example.backend.global.exception.ErrorCode
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * 어항 로그 서비스
 * 어항의 환경 데이터(온도, pH 등)를 관리하는 서비스
 */
@Service
@Transactional(readOnly = true)
class AquariumLogService(
    private val aquariumLogRepository: AquariumLogRepository,
    private val aquariumRepository: AquariumRepository
) : AbstractLogService<AquariumLog, AquariumLogRequestDto, AquariumLogResponseDto, Aquarium>() {

    override fun getLogRepository(): JpaRepository<AquariumLog, Long> = aquariumLogRepository

    override fun getParentRepository(): JpaRepository<Aquarium, Long> = aquariumRepository

    override fun getLogNotFoundErrorCode(): ErrorCode = ErrorCode.AQUARIUM_LOG_NOT_FOUND

    override fun getParentNotFoundErrorCode(): ErrorCode = ErrorCode.AQUARIUM_NOT_FOUND

    override fun createEntity(requestDto: AquariumLogRequestDto, aquarium: Aquarium): AquariumLog {
        // @NotNull validation이 통과했으므로 null이 아님
        val temperature = requestDto.temperature 
            ?: throw BusinessException(ErrorCode.INVALID_INPUT_VALUE)
        val ph = requestDto.ph 
            ?: throw BusinessException(ErrorCode.INVALID_INPUT_VALUE)
        val logDate = requestDto.logDate 
            ?: throw BusinessException(ErrorCode.INVALID_INPUT_VALUE)
        
        return AquariumLog(
            aquarium = aquarium,
            temperature = temperature,
            ph = ph,
            logDate = logDate
        )
    }

    override fun convertToResponseDto(entity: AquariumLog): AquariumLogResponseDto {
        return AquariumLogResponseDto.from(entity)
    }

    override fun updateEntity(entity: AquariumLog, requestDto: AquariumLogRequestDto, aquarium: Aquarium) {
        // @NotNull validation이 통과했으므로 null이 아님
        val temperature = requestDto.temperature 
            ?: throw BusinessException(ErrorCode.INVALID_INPUT_VALUE)
        val ph = requestDto.ph 
            ?: throw BusinessException(ErrorCode.INVALID_INPUT_VALUE)
        val logDate = requestDto.logDate 
            ?: throw BusinessException(ErrorCode.INVALID_INPUT_VALUE)
        
        entity.aquarium = aquarium
        entity.temperature = temperature
        entity.ph = ph
        entity.logDate = logDate
    }

    override fun findByParentId(parentId: Long): List<AquariumLog> {
        return aquariumLogRepository.findByAquariumId(parentId)
    }

    /**
     * 어항 로그 생성
     * @param requestDto 어항 로그 요청 DTO
     * @return 생성된 어항 로그 응답 DTO
     */
    @Transactional
    fun createLog(requestDto: AquariumLogRequestDto): AquariumLogResponseDto {
        val aquariumId = requestDto.aquariumId 
            ?: throw BusinessException(ErrorCode.INVALID_INPUT_VALUE)
        return createLog(requestDto, aquariumId)
    }

    /**
     * 어항 ID로 로그 목록 조회
     * @param aquariumId 어항 ID
     * @return 어항 로그 응답 DTO 목록
     */
    fun getLogsByAquariumId(aquariumId: Long): List<AquariumLogResponseDto> {
        return getLogsByParentId(aquariumId)
    }

    /**
     * 어항 로그 수정
     * @param logId 로그 ID
     * @param requestDto 어항 로그 요청 DTO
     * @return 수정된 어항 로그 응답 DTO
     */
    @Transactional
    fun updateLog(logId: Long, requestDto: AquariumLogRequestDto): AquariumLogResponseDto {
        val aquariumId = requestDto.aquariumId 
            ?: throw BusinessException(ErrorCode.INVALID_INPUT_VALUE)
        return updateLog(logId, requestDto, aquariumId)
    }

    /**
     * 어항 로그 삭제
     * @param logId 로그 ID
     */
    @Transactional
    override fun deleteLog(logId: Long) {
        val entity = aquariumLogRepository.findById(logId)
            .orElseThrow { BusinessException(ErrorCode.AQUARIUM_LOG_NOT_FOUND) }
        aquariumLogRepository.delete(entity)
    }
}
